qualtran/bloqs/arithmetic/negate.py

```diff
@@ -0,0 +1,74 @@
+#  Copyright 2024 Google LLC
+#
+#  Licensed under the Apache License, Version 2.0 (the "License");
+#  you may not use this file except in compliance with the License.
+#  You may obtain a copy of the License at
+#
+#      https://www.apache.org/licenses/LICENSE-2.0
+#
+#  Unless required by applicable law or agreed to in writing, software
+#  distributed under the License is distributed on an "AS IS" BASIS,
+#  WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+#  See the License for the specific language governing permissions and
+#  limitations under the License.
+from functools import cached_property
+from typing import TYPE_CHECKING
+
+from attrs import frozen
+
+from qualtran import Bloq, bloq_example, BloqDocSpec, QDType, QInt, Signature
+from qualtran.bloqs.arithmetic import AddK
+from qualtran.bloqs.basic_gates import OnEach, XGate
+
+if TYPE_CHECKING:
+    from qualtran import BloqBuilder, SoquetT
+
+
+@frozen
+class Negate(Bloq):
+    """Compute the two's complement for a signed integer/fixed-point value.
+
+    The two's complement is computed by the following steps:
+    1. Flip all the bits
+    2. Add 1 to the value (interpreted as an unsigned integer), ignoring the overflow.
+
+    For a controlled negate bloq: the second step uses a quantum-quantum adder by
+    loading the constant (i.e. 1), therefore has an improved controlled version
+    which only controls the constant load and not the adder circuit, hence halving
+    the T-cost compared to a controlled adder.
+
+    Args:
+        dtype: The data type of the input value.
+
+    Registers:
+        x: Any signed value stored in two's complement form.
+    """
+
+    dtype: QDType
```

Comment:
Should we assert that `dtype` is a signed type ? So `QInt` or `QFxp(signed=True)` ?

Reply:
It actually works for unsigned types as well, it's the [unary minus operator in c++](https://en.cppreference.com/w/cpp/language/operator_arithmetic). Perhaps we can use the c++ naming convention?

Reply:
Thanks, we should at least have a link explaining the behavior for unsigned types in the docstring. I don't have strong opinions on naming convention.

Reply:
updated docstring to explain this

Reply:
will leave the Bloq as `Negate` for now